internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -314,6 +327,31 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 	return cb, nil
 }
 
+func waitForWatcher(ctx context.Context, log *logger.Logger, markerFilePath string, waitTime time.Duration) error {
+	//Wait for the watcher to be up and running
+	watcherContext, cancel := context.WithTimeout(ctx, waitTime)
+	defer cancel()
+
+	markerWatcher := newMarkerFileWatcher(markerFilePath, log)
+	err := markerWatcher.Run(watcherContext)
+	if err != nil {
+		return fmt.Errorf("error starting update marker watcher: %w", err)
+	}
+
+	for {
+		select {
+		case updMarker := <-markerWatcher.Watch():
```

Comment:
We are also receiving from the channel returned by `markerWatcher.Watch()` in the coordinator: 

https://github.com/elastic/elastic-agent/blob/d8f513438813e9faaf3d3f86823e3f1593178b55/internal/pkg/agent/application/coordinator/coordinator.go#L403-L405

https://github.com/elastic/elastic-agent/blob/d8f513438813e9faaf3d3f86823e3f1593178b55/internal/pkg/agent/application/coordinator/coordinator.go#L1020-L1023

I wonder if we might encounter a race condition, depending on which receive happens first?

Reply:
The reason why I use a new update marker watcher is exactly not to "steal" events from coordinator and avoid race conditions.

I am relying on inotify implementation to send the correct events to both watchers here so I don't need to rebroadcast the WATCHING state

Reply:
Oh duh, I missed the constructor call.  All good then 🙂.